File: solutions/go/resistor-color-trio/1/resistor_color_trio.go
package resistorcolortrio

import (
    "slices"
    "strings"
)

var Colors = []string{
    "black",
    "brown",
    "red",
    "orange",
    "yellow",
    "green",
    "blue",
    "violet",
    "grey",
    "white",
}

var Quantifiers = []string{
    "ohms",
    "kiloohms",
    "megaohms",
    "gigaohms",
}

func Label(colors []string) string {
	n1 := slices.Index(Colors, colors[0])
    n2 := slices.Index(Colors, colors[1])
    n3 := slices.Index(Colors, colors[2])
    var additionalDigits int
    if n1 != 0 {
        additionalDigits += 2
    } else if n2 != 0 {
        additionalDigits++
    }
    var quantifier int
    if n3 + additionalDigits > 3 {
        quantifier++
    }
    if n3 + additionalDigits > 6 {
        quantifier++
    }
    if n3 + additionalDigits > 9 {
        quantifier++
    }
    if additionalDigits == 0 {
        quantifier = 0
    }
    nDigits := n3 + additionalDigits - 3 * quantifier
    var builder strings.Builder
    if n1 != 0 {
        builder.WriteRune('0' + rune(n1))
    }
    if n1 == 0 || n1 != 0 && nDigits >= 2 {
        builder.WriteRune('0' + rune(n2))
    }
    if nDigits == 3 {
        builder.WriteRune('0')
    }
    builder.WriteRune(' ')
    builder.WriteString(Quantifiers[quantifier])
    return builder.String()
}

File: solutions/java/resistor-color-trio/1/src/main/java/ResistorColorTrio.java
import java.util.Arrays;
import java.util.List;

class ResistorColor {
    static final private List<String> COLORS = List.of(
        new String[]{
            "black", "brown", "red", "orange", "yellow", "green", "blue", "violet", "grey", "white"
        }
    );
    
    static int colorCode(String color) {
        return COLORS.indexOf(color);
    }

    String[] colors() {
        return COLORS.toArray(new String[0]);
    }
}

class ResistorColorDuo {
    static int value(String[] colors) {
        if (colors.length < 2) {
            throw new IllegalArgumentException("Less than 2 colors: " + colors.toString());
        }
        int first = ResistorColor.colorCode(colors[0]);
        int second = ResistorColor.colorCode(colors[1]);
        if (first != -1 && second != -1) {
            return first * 10 + second;
        }
        if (first == -1) {
            throw new IllegalArgumentException("Invalid first color: " + colors[0]);
        }
        throw new IllegalArgumentException("Invalid second color: " + colors[1]);
    }
}

class ResistorColorTrio {
    static final private List<String> PREFIXES = List.of(
        new String[]{
            "ohms", "kiloohms", "megaohms", "gigaohms"
        }
    );
    
    static String label(String[] colors) {
        if (colors.length < 3) {
            throw new IllegalArgumentException("Less than 3 colors: " + colors.toString());
        }
        int value = ResistorColorDuo.value(colors);
        int zeros = ResistorColor.colorCode(colors[2]);
        if (zeros == -1) {
            throw new IllegalArgumentException("Invalid third color: " + colors[2]);
        }
        String valueString = Integer.toString(value);
        String numberString = valueString + "0".repeat(zeros);
        int unitIndex = (numberString.length() - 1) / 3;
        String unit = PREFIXES.get(unitIndex);
        int digitLength = numberString.length() - unitIndex * 3;
        String digitString = numberString.substring(0, digitLength);
        String fractionString = numberString.substring(digitLength).replaceAll("0+$", "");
        if (fractionString.length() > 0) {
            digitString += "." + fractionString;
        }
        return digitString + " " + unit;
    }
}
